### PR TITLE
Merge THM Module into MOOSE

### PR DESCRIPTION
**Pull request by permcody**
## Reason

Merging the THM repository into the MOOSE repository. It has been approved for OSS release

## Design

See the THM design documents and SDR.

## Impact

This PR adds a completely new physics module (Major)

### CODE REVIEW

**Comment by loganharbour** [comment]
- Remove `.coverage` and add to the root `.coverage` file with `[thm]` instead of `[default`]
- Remove .gitignore

**Comment by loganharbour** [comment]
Script used for porting issues in case we need it later:

```python
import gitlab
from github import Github
import time
from datetime import datetime, timedelta
import re
import csv

gitlab_url = 'https://gitlab.software.inl.gov'
gitlab_token = ''
github_token = ''
repo = 'idaholab/thm'
to_repo = 'idaholab/moose'

gl = gitlab.Gitlab(gitlab_url, private_token=gitlab_token)
gh = Github(github_token)

gl_repo = gl.projects.get(repo)
gh_repo = gh.get_repo(to_repo)

# To be filled with old issue ID -> new issue ID
issue_map = {}

# Name map for the header from person -> github persona
author_map = {'Alexander Lindsay': 'lindsayad',
            'David Andrs': 'andrsd',
            'Fande Kong': 'fdkong',
            'Guillaume Giudicelli': 'GiudGiud',
            'Joshua Hansel': 'joshuahansel',
            'Logan Harbour': 'loganharbour',
            'Sebastian Schunert': 'snschune'}

# User map from gitlab -> github persona
user_map = {'hansje': 'joshuahansel',
           'balep': 'pbalest',
           'schuseba': 'snschune',
           'schus': 'snschune',
           'marcm': 'rcm59',
           'KONGF': 'fdkong',
           'lindad': 'lindsayad',
           'ortej': 'jortensi',
           'LABOVM': 'vincentlaboure',
           'rcm': 'rcm59',
           'slauae': 'aeslaughter'}

# Converts a gitlab time -> a human time (this is dirty but it works...)
def human_time(timestamp):
    created = datetime.strptime(timestamp[:-5], '%Y-%m-%dT%H:%M:%S') - timedelta(hours=7)
    return created.strftime('%b %d, %Y %I:%M%p MST')

# Replaces an author with a name to their github username
def replace_author(object):
    if object.author['name'] in author_map:
        return '@' + author_map[object.author['name']]
    return object.author['name']

# Replaces @ mentions
def replace_mentions(str):
    for from, to in user_map.items():
        str = str.replace('@' + from, '@' + to)
    return str

# Header for adding a comment for an issue/note/etc
def body_header(issue, object, extra_content=None):
    new_body = 'Imported from [{}#{}]({}) '.format(repo, issue.iid, issue.web_url)
    new_body += 'by {} at {}\n\n'.format(replace_author(object), human_time(object.created_at))
    if extra_content:
        new_body += extra_content + '\n\n'
    new_body += '------\n\n'
    return new_body

# Gets a URL to a merge request on a gitlab instance
def gitlab_merge_request_url(num):
    return gitlab_url + '/' + repo + '/-/merge_requests/{}'.format(num)

# Helper for calling a github action that can be rate limited; will try it every 30 sec
# until the rate liminting is gone
def do_github(action):
    while True:
        try:
            result = action()
            time.sleep(30)
            return result
        except:
            print('Rate limited')
            time.sleep(30)

# Issues that have uploads that we need to manually change
has_uploads = set()

issues = sorted(gl_repo.issues.list(all=True), key=lambda entry: entry.iid)
print('Importing {} issues...'.format(len(issues)))
for i in range(len(issues)):
    issue = issues[i]
    print('{}/{}: !{} "{}"'.format(i + 1, len(issues), issue.iid, issue.title))

    # Whether or not the issue was closed by a merge request
    # (to trigger a MR closed comment instead of a generic comment)
    closed_by_merge_request = None
    # The merge requests that were mentioned in the issue
    mentioned_merge_requests = set()
    # Comments we need to add
    comments = []

    # Work through all of the notes, which could be mentions or actual user comments
    for note in sorted(issue.notes.list(all=True), key=lambda entry: datetime.strptime(entry.created_at[:-5], '%Y-%m-%dT%H:%M:%S').timestamp()):
        # System notes are mentions (MRs, commits, etc)
        if note.system:
            search_closed_merge_request = re.search('closed via merge request !([0-9]+)', note.body)
            if search_closed_merge_request:
                mr_num = search_closed_merge_request.group(1)
                closed_by_merge_request = mr_num
                mentioned_merge_requests.add(mr_num)

            search_mentioned_merge_request = re.search('mentioned in merge request !([0-9]+)', note.body)
            if search_mentioned_merge_request:
                mr_num = search_mentioned_merge_request.group(1)
                mentioned_merge_requests.add(mr_num)
        # Standard comment by a user
        else:
            # Keep track of issues where we need to manually add uploads
            if '/uploads' in note.body:
                has_uploads.add(issue.iid)

            comment_body = body_header(issue, note) + replace_mentions(note.body)
            comments.append(comment_body)

    # If not closed by a MR but it was closed, add a comment as to when it was closed
    if issue.state == 'closed' and closed_by_merge_request is None:
        comments.append('Closed at {}.'.format(human_time(issue.closed_at)))

    # Create the extra content in the header for mentioned MRs
    extra_content = None
    if mentioned_merge_requests:
        extra_content = 'Mentioned merge requests:\n\n'
        for mr_num in sorted(mentioned_merge_requests):
            mr = gl_repo.mergerequests.get(mr_num)
            link = '[{} (!{})]({})'.format(mr.title, mr_num, gitlab_merge_request_url(mr_num))
            extra_content += '- ' + link + '\n'
            if mr_num == closed_by_merge_request:
                comments.append('Closed by merge request ' + link + ' at ' + human_time(mr.merged_at))

    # Keep track of issues where we need to manually add uploads
    if '/uploads' in issue.description:
        has_uploads.add(issue.iid)

    # Create the issue
    issue_body = body_header(issue, issue, extra_content) + replace_mentions(issue.description)
    create_issue = lambda: gh_repo.create_issue(title='THM: ' + issue.title, body=issue_body, labels=['C: Modules'])
    new_issue = do_github(create_issue)

    # Save the new issue number to the old -> new ID map
    with open('issue_map.csv', 'a') as f:
        csv.writer(f).writerow((issue.iid, new_issue.number))

    # Post all of the comments
    for comment in comments:
        do_github(lambda: new_issue.create_comment(comment))

    # Close the issue if it's closed
    if issue.state == 'closed':
        do_github(lambda: new_issue.edit(state='closed'))

if has_uploads:
    print('\nThe following issue(s) have uploads and will need to be edited manually:')
    for issue_id in sorted(has_uploads):
        issue = gl_repo.issues.get(issue_id)
        print('- !{} "{}"'.format(issue_id, issue.title))
```

File: modules/modules.mk [hunk 1]
```diff
@@ -31,55 +31,60 @@ ifeq ($(ALL_MODULES),yes)
...
 
-ifeq ($(XFEM),yes)
-        TENSOR_MECHANICS            := yes
+ifeq ($(NAVIER_STOKES),yes)
+        FLUID_PROPERTIES            := yes
+        HEAT_CONDUCTION             := yes
+        RDG                         := yes
 endif
 
-# Modules that follow have one or more dependencies
```
Comment (loganharbour) [inline modules/modules.mk:64]: Why remove this comment? Breaking the point where the includes aren't independent anymore I think is good

File: modules/thm/include/base/THMApp.h [hunk 2]
```diff
@@ -65,6 +65,13 @@ class THMApp : public MooseApp
   static void registerApps();
   static void registerAll(Factory & f, ActionFactory & af, Syntax & s);
 
+  /**
+   * Deprecated Methods
+   */
+  static void registerObjects(Factory & factory);
+  static void associateSyntax(Syntax & syntax, ActionFactory & action_factory);
+  static void registerExecFlags(Factory & factory);
+
```
Comment (loganharbour) [inline modules/thm/include/base/THMApp.h:74]: I say don't even include these if we didn't have them in THM before

**Review by loganharbour — CHANGES_REQUESTED** [review]
Remove

- `modules/thm/.clang-format`
- `modules/thm/.coverage`
- `modules/thm/.gitlab`
- `modules/thm/.mailmap`

**Comment by lindsayad** [comment]
Wow @andrsd is going to build an insurmountable commit count lead in MOOSE once this is merged 😄 ~A question I would ask is whether it's universally known that "thermal hydraulics" means essentially system thermal hydraulics. I feel like the answer is probably yes, but if it wasn't then this could be confusing to users that we now have a thermal hydraulics module that is separate from things like `heat_conduction` (whose name should probably be updated to be more like `heat_transfer`) and `navier_stokes`.~ Belay my concerns. I see that THM already pulls in those aforementioned modules, so I think it's fair to consider this to be a cohesive (albeit mostly 1D) thermal hydraulics module.

**Review by cticenhour — DISMISSED** [review]
Since this is likely to be merged in before the end of the month, I wanted to add a general review/reminder to add a note about THM to the January newsletter now that `modules/doc/content/newsletter/2022_01.md` has made it to devel. Thanks!

**Comment by lindsayad** [comment]
Any rough timeline on when we think this might be merged?

**Comment by andrsd** [comment]
I resolved the doco issue yesterday, @loganharbour is still looking into some coverage issues. After that is resolved, we can merge. I think.

**Comment by fdkong** [comment]
I would like to see this being merged. What can I help?

**Comment by permcody** [comment]
Should we tell him @loganharbour? 😄

**Comment by fdkong** [comment]
3K commits

**Comment by loganharbour** [comment]
Something really weird is going on now that we've added this module. 

See the "Modules" recipe (which tests modules individually for coverage): https://civet.inl.gov/job/937847/ (6 hours) vs the current run on next https://civet.inl.gov/job/939598/ (1 hour). The difference is in the extraction of coverage. This went from taking < 20s per module to > 12m per module.

In specific, the final extraction (taking `chemical_reactions`) as an example:

```
BUILD_ROOT/moose/: lcov --gcov-tool gcov --extract /ramdisk/civet/build_0/moose_coverage/final.info \*/ramdisk/civet/build_0/moose/modules/chemical_reactions/src/\* --extract /ramdisk/civet/build_0/moose_coverage/final.info \*/ramdisk/civet/build_0/moose/modules/chemical_reactions/include/\* --output-file /ramdisk/civet/build_0/moose_coverage/coverage.info
Reading tracefile /ramdisk/civet/build_0/moose_coverage/final.info
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/include/auxkernels/KineticDisPreConcAux.h
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/include/auxkernels/KineticDisPreRateAux.h
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/include/utils/EquilibriumConstantFit.h
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/actions/AddCoupledEqSpeciesAction.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/actions/AddCoupledSolidKinSpeciesAction.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/actions/AddPrimarySpeciesAction.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/actions/AddSecondarySpeciesAction.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/auxkernels/AqueousEquilibriumRxnAux.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/auxkernels/EquilibriumConstantAux.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/auxkernels/KineticDisPreConcAux.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/auxkernels/KineticDisPreRateAux.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/auxkernels/PHAux.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/auxkernels/TotalConcentrationAux.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/base/ChemicalReactionsApp.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/bcs/ChemicalOutFlowBC.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/kernels/CoupledBEEquilibriumSub.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/kernels/CoupledBEKinetic.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/kernels/CoupledConvectionReactionSub.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/kernels/CoupledDiffusionReactionSub.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/kernels/DarcyFluxPressure.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/kernels/DesorptionFromMatrix.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/kernels/DesorptionToPorespace.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/kernels/PrimaryConvection.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/kernels/PrimaryDiffusion.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/kernels/PrimaryTimeDerivative.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/main.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/materials/LangmuirMaterial.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/materials/MollifiedLangmuirMaterial.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/postprocessors/TotalMineralVolumeFraction.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/utils/EquilibriumConstantFit.C
Extracted 30 files
Writing data to /ramdisk/civet/build_0/moose_coverage/coverage.info
Summary coverage rate:
  lines......: 93.3% (873 of 936 lines)
  functions..: 89.7% (96 of 107 functions)
  branches...: no data found
```

Is what takes many minutes. It will print out the summary and then hang. I have absolutely no clue what is going on here and I don't have the immediate time to debug it. Any takers? This should be able to be replicated on rod with:

```
moosebuild
module load mpich cmake lcov python vtk
git clone https://github.com/permcody/moose
cd moose
git checkout thm_merge_fixes_2
MOOSE_JOBS=32 ./scripts/update_and_rebuild_petsc.sh
METHODS=opt MOOSE_JOBS=32 ./scripts/update_and_rebuild_libmesh.sh
MOOSE_DIR=$(pwd)
cd framework
make -j 32
cd ../modules/chemical_reactions
MOOSE_HEADER_SYMLINKS=false make -j 32 coverage=true graceful=true
lcov --gcov-tool gcov --capture --initial --directory $MOOSE_DIR/modules/chemical_reactions --output-file $MOOSE_DIR/initialize.info
./run_tests -j 32
lcov --gcov-tool gcov --capture --ignore-errors gcov,source --directory $MOOSE_DIR/modules/chemical_reactions --output-file $MOOSE_DIR/covered.info
lcov --gcov-tool gcov --add-tracefile $MOOSE_DIR/initialize.info --add-tracefile $MOOSE_DIR/covered.info --output-file $MOOSE_DIR/final.info
lcov --gcov-tool gcov --extract $MOOSE_DIR/final.info $MOOSE_DIR/modules/chemical_reactions/src/* --extract $MOOSE_DIR/final.info $MOOSE_DIR/modules/chemical_reactions/include/* --output-file $MOOSE_DIR/coverage.info
```

The last command should hang. I typed all of the above without testing... it _should_ work

**Review by loganharbour — DISMISSED** [review]

**Review by andrsd — DISMISSED** [review]

**Review by loganharbour — APPROVED** [review]

**Review by andrsd — APPROVED** [review]

**Pull request merged**